lib/ansible/plugins/filter/core.py

```diff
@@ -253,42 +254,19 @@ def get_hash(data, hashtype='sha1'):
     return h.hexdigest()
 
 
-def get_encrypted_password(password, hashtype='sha512', salt=None):
+def get_encrypted_password(password, hashtype='sha512', salt=None, **settings):
+    if salt is not None:
+        settings['salt'] = salt
 
-    # TODO: find a way to construct dynamically from system
-    cryptmethod = {
-        'md5': '1',
-        'blowfish': '2a',
-        'sha256': '5',
-        'sha512': '6',
+    passlib_mapping = {
+        'md5': 'md5_crypt',
+        'blowfish': 'bcrypt',
+        'sha256': 'sha256_crypt',
+        'sha512': 'sha512_crypt',
     }
 
-    if hashtype in cryptmethod:
-        if salt is None:
-            r = SystemRandom()
-            if hashtype in ['md5']:
-                saltsize = 8
-            else:
-                saltsize = 16
-            saltcharset = string.ascii_letters + string.digits + '/.'
-            salt = ''.join([r.choice(saltcharset) for _ in range(saltsize)])
-
-        if not HAS_PASSLIB:
-            if sys.platform.startswith('darwin'):
-                raise AnsibleFilterError('|password_hash requires the passlib python module to generate password hashes on Mac OS X/Darwin')
-            saltstring = "$%s$%s" % (cryptmethod[hashtype], salt)
-            encrypted = crypt.crypt(password, saltstring)
-        else:
-            if hashtype == 'blowfish':
-                cls = passlib.hash.bcrypt
-            else:
-                cls = getattr(passlib.hash, '%s_crypt' % hashtype)
-
-            encrypted = cls.encrypt(password, salt=salt)
-
-        return encrypted
-
-    return None
+    hashtype = passlib_mapping.get(hashtype, hashtype)
+    return passlib_or_crypt(password, hashtype, error_type=AnsibleFilterError, **settings)
```

Comment:
Because I think otherwise there could be unexpected behaviour. A typo in the hash-type or a unsupported type by crypt should not lead to the function being successful. Instead the issue should be pointed out right away.
This also makes the function more consistent with the old do_encrypt.py of encrypt.py.

Reply:
Typically, our code uses try: except to catch and transform exception rather than making explicit passing of exception types part of a function's signature.  There are drawbacks to try: except (for instance, changing the traceback on python2) but I think we should stick with that idiom.

Reply:
I modified it to always raise AnsibleError.
So also in the filter context (password_hash).
Please tell me whether I should raise AnsibleFilterError there (via the steps you suggested -> https://nedbatchelder.com/blog/200711/rethrowing_exceptions_in_python.html)?